### PR TITLE
Use individual section files #137 #184

### PR DESCRIPTION
**Pull request by wilzbach**
@stonemaster I found a bit of time and gave #137 (the split-up in folders) a try.

Notable changes
-----------------------

- I use `index.yml` files as
    - renaming will be a pain
    - renames will mess up the git history (very problematic for translations)
    - we need a config file for the chapter title anyways
- a file can now only contain one `h1` element -> it will be it's title (that's a restriction for normal markdown/html too btw)
- `h2` sections are now allowed - I didn't see a reason anymore for this limitation and it will work smoother with markdown editor (e.g. Github)
- file recognition is now only semi-automatic (it's based on the index.yml ordering) - as mentioned before I believe that's the best way and most transparent way to solve this
- As a proof-of-concept I only splitted the welcome tour to avoid merge conflicts
- I didn't make further changes, but once this is merged we can start adding more cool stuff like "Edit this section on Github"
- I tried to keep the changes at a minimum, but for example a next (easy) step would be to have the tour start page and title in the `index.yml` too

Cheers,

Seb

### CODE REVIEW

File: source/contentprovider.d [hunk 1]
```diff
@@ -78,56 +72,70 @@ class ContentProvider
 			auto parts = filename[contentDirectory.length .. $]
 					.split('/').filter!(x => !x.empty)
 					.array;
-			logInfo("Found content file '%s'", parts);
+			// search for language-specific root file
```
Comment (stonemaster) [inline source/contentprovider.d]: indentation

File: source/contentprovider.d [hunk 2]
```diff
@@ -78,56 +72,70 @@ class ContentProvider
...
-					enforce(currentSection != 0, new Exception("%s: level 3 section can't be first (%s)".format(filename, section.title)));
-					auto content = updateContent(language, chapter, currentSection);
-					content.html ~= filterMarkdown(section.content,
-							MarkdownFlags.backtickCodeBlocks | MarkdownFlags.vanillaMarkdown);
-				} else {
-					throw new Exception("%s: Illegal section %s".format(filename, section.title));
-				}
+	private void addChapter(string chapter, string chapterDir, string language)
+	{
+	    auto configFile = buildPath(chapterDir, "index.yml");
```
Comment (stonemaster) [inline source/contentprovider.d]: indent

File: source/contentprovider.d [hunk 3]
```diff
@@ -78,56 +72,70 @@ class ContentProvider
...
+            auto filename = buildPath(chapterDir, section ~ ".md");
+            Content* content = addSection(filename, chapter, section, language);
+            content._id = i++;
+        }
+	}
+
+	private Content* addSection(string filename, string chapter, string currentSection, string language)
+	{
+	    Content* content = updateContent(language, chapter, currentSection);
+        enforce(exists(filename), "couldn't find " ~ filename);
```
Comment (stonemaster) [inline source/contentprovider.d]: indent

**Comment by stonemaster** [comment]
This PR can't be merged at the moment right? Because wondering how we proceed.
1. Split up all files
2. Have both systems - old & new supported and convert step by step.

**Comment by stonemaster** [comment]
Btw. I like the `index.yml` format.

File: source/contentprovider.d [hunk 4]
```diff
@@ -192,34 +200,30 @@ class ContentProvider
...
-			chapter.sections.length = sections.length;
-			foreach(sectionIdx, ref content; sections) {
-				auto section = &chapter.sections[sectionIdx - 1];
-				section.title = content.title;
-				section.sectionId = sectionIdx;
-			}
-		}
-
-		return toc;
+        struct Chapter {
```
Comment (stonemaster) [inline source/contentprovider.d]: indent

**Comment by wilzbach** [comment]
The indentation is correct, probably just the usual problem that my vim is configured to use whitespaces. I will fix that later. 

For preceeding I would just write a small script that semi-automates the conversion. Imho that's less work than supporting both versions somehow.

File: source/contentprovider.d [hunk 5]
```diff
@@ -192,34 +200,30 @@ class ContentProvider
...
+            string title;
+            string chapterId;
+            Tuple!(string, "title", string, "sectionId")[] sections;
+        }
+	    Chapter[] toc;
+        auto chapterTitles = language in chapterTitle_;
+        enforce(chapterTitles !is null, new Exception("%s not known.".format(language)));
+
+        // files are already sorted
+        foreach (chapterId, sections; content_[language]) {
```
Comment (stonemaster) [inline source/contentprovider.d]: Is this really sorted? _content is a hash map and so doesn't guarantee any ordering.

**Comment by stonemaster** [comment]
>The indentation is correct, probably just the usual problem that my vim is configured to use whitespaces. I will fix that later.
> For preceeding I would just write a small script that semi-automates the conversion. Imho that's less work than supporting both versions somehow.

Anothe possibility is to just add a vim modeline so vim will pick up the right settings for that file.

File: source/contentprovider.d (hunk 5 above)
Comment (wilzbach) [inline source/contentprovider.d]: Oh yeah I have to remove the comment - for the ordering a `._id` property is now created in the setup above

File: source/webinterface.d [hunk 6]
```diff
@@ -125,20 +129,21 @@ class WebInterface
...
-	void getTour(HTTPServerRequest req, HTTPServerResponse res, string _chapter, int _section)
+	void getTour(HTTPServerRequest req, HTTPServerResponse res, string _chapter, string _section)
 	{
 		auto sec = getTourDataAndValidate(_chapter, _section);
 		auto htmlContent = sec.tourData.content.html;
 		auto chapterId = _chapter;
 		auto hasSourceCode = !sec.tourData.content.sourceCode.empty;
 		auto sourceCodeEnabled = sec.tourData.content.sourceCodeEnabled;
 		auto section = _section;
+		auto sectionId =  sec.tourData.content._id;
```
Comment (stonemaster) [inline source/webinterface.d]: Is sectionId still needed? Shouldn't it be replaces by the short title?

**Comment by wilzbach** [comment]
> Anothe possibility is to just add a vim modeline so vim will pick up the right settings for that file.

Indention/tab problem is fixed - sorry for the troubles.

**Comment by stonemaster** [comment]
One remaining question: The old link format CHAPTER/ID is obselete then right? What about old bookmarks? Should we support them?

**Comment by stonemaster** [comment]
How would htaccess work when we just use vibe.d? 
A route seems like the most sensible option. Does vibe.d support overloaded paths? Currently we match for :chapter/:section and the most easiest fix would be to have a :chapter/:section overload that parses :section as an integer... 
If this compatibility feature just ends being too much work I am also okay with just dropping old format's support. The tour isn't online for too long so I think we could live with a breaking change at this stage.

**Comment by wilzbach** [comment]
> How would htaccess work when we just use vibe.d? 

oh I thought you use nginx or apache as reverse proxy.

> A route seems like the most sensible option. Does vibe.d support overloaded paths? Currently we match for :chapter/:section and the most easiest fix would be to have a :chapter/:section overload that parses :section as an integer... 

I don't know about path overloading, but we can just use `isNumeric` - I added this as a new commit. 
We probably want to send a 301 permanent `redirect` response, but that can be improved in the future :)

**Comment by stonemaster** [comment]
Nice idea with `isNumeric`. If everything passes I'm okay with merging.
Just let's double check after this has been built and deployed that everything is working as before

**Comment by wilzbach** [comment]
> If everything passes I'm okay with merging.

it's green for me :)

> Just let's double check after this has been built and deployed that everything is working as before

Alrighty!

**Comment by wilzbach** [comment]
@stonemaster I rebased according to 049259dde23b4b40ca1923c8d8571151ce5f02db.
Can I merge this once Travis is green?

**Comment by stonemaster** [comment]
Let's make this happen! 👍

**Comment by wilzbach** [comment]
Let's rock then :)

**Comment by wilzbach** [comment]
> Just let's double check after this has been built and deployed that everything is working as before

Everything looks ok on my end - I started to rebase the other pending PRs ;-)

**Comment by stonemaster** [comment]
Great work, thanks! Did some dumb testing and everything seems to run just fine :-)

**Pull request merged**